exchanges/poloniex/poloniex_wrapper.go

```diff
@@ -491,259 +665,585 @@ func (e *Exchange) SubmitOrder(ctx context.Context, s *order.Submit) (*order.Sub
+		return nil, fmt.Errorf("%w: %q", asset.ErrNotSupported, action.AssetType)
+	}
+	switch action.Type {
+	case order.Market, order.Limit, order.LimitMaker:
+		resp, err := e.CancelReplaceOrder(ctx, &CancelReplaceOrderRequest{
+			OrderID:           action.OrderID,
+			ClientOrderID:     action.ClientOrderID,
+			Price:             action.Price,
+			Quantity:          action.Amount,
+			AmendedType:       action.Type.String(),
+			TimeInForce:       TimeInForce(action.TimeInForce),
+			SlippageTolerance: 0,
+		})
+		if err != nil {
+			return nil, err
+		}
+		modResp, err := action.DeriveModifyResponse()
+		if err != nil {
+			return nil, err
+		}
+		modResp.OrderID = resp.ID
+		return modResp, nil
+	case order.Stop, order.StopLimit:
+		oResp, err := e.CancelReplaceSmartOrder(ctx, &CancelReplaceSmartOrderRequest{
+			OrderID:          action.OrderID,
+			ClientOrderID:    action.ClientOrderID,
+			Price:            action.Price,
+			StopPrice:        action.TriggerPrice,
+			Amount:           action.Amount,
+			AmendedType:      OrderType(action.Type),
+			ProceedOnFailure: !action.TimeInForce.Is(order.ImmediateOrCancel),
+			TimeInForce:      TimeInForce(action.TimeInForce),
+		})
+		if err != nil {
+			return nil, err
+		}
+		modResp, err := action.DeriveModifyResponse()
+		if err != nil {
+			return nil, err
+		}
+		modResp.OrderID = oResp.ID
+		return modResp, nil
+	default:
+		return nil, fmt.Errorf("%w: %q", order.ErrUnsupportedOrderType, action.Type)
 	}
-	modResp.OrderID = strconv.FormatInt(resp.OrderNumber, 10)
-	return modResp, nil
 }
 
 // CancelOrder cancels an order by its corresponding ID number
 func (e *Exchange) CancelOrder(ctx context.Context, o *order.Cancel) error {
-	if err := o.Validate(o.StandardCancel()); err != nil {
+	if err := o.Validate(); err != nil {
 		return err
 	}
-
-	orderIDInt, err := strconv.ParseInt(o.OrderID, 10, 64)
-	if err != nil {
-		return err
+	if o.OrderID == "" && o.ClientOrderID == "" {
+		return order.ErrOrderIDNotSet
 	}
-
-	return e.CancelExistingOrder(ctx, orderIDInt)
-}
+	var err error
+	switch o.AssetType {
+	case asset.Spot:
+		switch o.Type {
+		case order.Stop, order.StopLimit, order.TrailingStop, order.TrailingStopLimit:
+			_, err = e.CancelSmartOrderByID(ctx, o.OrderID, o.ClientOrderID)
+		default:
+			_, err = e.CancelOrderByID(ctx, o.OrderID)
+		}
+	case asset.Futures:
+		_, err = e.CancelFuturesOrder(ctx, &CancelOrderRequest{Symbol: o.Pair.String(), OrderID: o.OrderID, ClientOrderID: o.ClientOrderID})
+	default:
+		return fmt.Errorf("%w: %q", asset.ErrNotSupported, o.AssetType)
+	}
+	return err
+}
 
 // CancelBatchOrders cancels an orders by their corresponding ID numbers
 func (e *Exchange) CancelBatchOrders(ctx context.Context, o []order.Cancel) (*order.CancelBatchResponse, error) {
 	if len(o) == 0 {
-		return nil, order.ErrCancelOrderIsNil
+		return nil, common.ErrEmptyParams
+	}
+	type assetAndType struct {
+		aType asset.Item
+		oType order.Type
+		pair  currency.Pair
```

Comment:
🚧 Change request:
I think future maintainers will want to understand why we have the logic we do below.
```suggestion
		pair  currency.Pair // only futures orders need grouping by pair
```